File: dotnet/CardsProject/Domain/Tournaments/Match.cs
using CardsProject.Domain.Players;
using System.ComponentModel.DataAnnotations.Schema;
using System.ComponentModel.DataAnnotations;
using System.Text.Json.Serialization;

namespace CardsProject.Domain.Tournaments;

public enum MatchStatusType
{
    Pending,
    Active,
    Completed,
    BYE,
    Draw
}

public class Match : IValidatableObject
{
    public int Id { get; set; }

    public int? TableNumber { get; set; } = null;
    public MatchStatusType Status { get; set; }
    public int Player1Wins { get; set; } = 0;
    public int Player2Wins { get; set; } = 0;
    [JsonPropertyName("startedAt")]
    public DateTime? StartedAt { get; set; } = null;
    [JsonPropertyName("endedAt")]
    public DateTime? EndedAt { get; set; } = null;
    public string? ResultNotes { get; set; }

    public int? RoundId { get; set; }
    [ForeignKey(nameof(RoundId))]
    public TournamentRound? Round { get; set; }
    public int? Player1Id { get; set; }
    [ForeignKey(nameof(Player1Id))]
    public Player? Player1 { get; set; }
    public int? Player2Id { get; set; }
    [ForeignKey(nameof(Player2Id))]
    public Player? Player2 { get; set; }

    // Business operations

    public void RecordResult(int p1Wins, int p2Wins)
    {
        // TODO: implement record_result
    }

    public void FinalizeResult()
    {
        // TODO: implement finalize_result
    }

    public bool DetermineWinner()
    {
        // TODO: implement determine_winner
        return default;
    }

    public void Concede(int playerId)
    {
        // TODO: implement concede
    }

    public void Draw()
    {
        // TODO: implement draw
    }

    // ── Lifecycle state machine ──────────────────────────────────────
    private static readonly System.Collections.Generic.Dictionary<MatchStatusType, MatchStatusType[]> AllowedTransitions = new()
    {
        [MatchStatusType.Pending] = new[] { MatchStatusType.Active, MatchStatusType.BYE },
        [MatchStatusType.Active] = new[] { MatchStatusType.Completed, MatchStatusType.Draw }
    };

    public void AssertTransition(MatchStatusType to)
    {
        if (!AllowedTransitions.TryGetValue(Status, out var allowed) || !System.Array.Exists(allowed, s => s == to))
            throw new InvalidOperationException($"Transition {Status} -> {to} not allowed");
    }

    // ── Domain invariants (simple rules) ──────────────────────────────
    public IEnumerable<ValidationResult> Validate(ValidationContext validationContext)
    {
        if (!( (Player1Wins >= 0 && Player2Wins >= 0) ))
            yield return new ValidationResult("Win counts must not be negative", new[] { nameof(Id) });
        if (!( (Player1Wins >= 0 && Player1Wins <= 2 && Player2Wins >= 0 && Player2Wins <= 2) ))
            yield return new ValidationResult("Win counts cannot exceed 2 in a best-of-3 match", new[] { nameof(Id) });
    }
}

File: rails/app/models/match.rb
class Match < ApplicationRecord
  self.table_name = 'matches'

  enum :status, { pending: 0, active: 1, completed: 2, b_y_e: 3, draw: 4 }

  belongs_to :round, class_name: 'TournamentRound'
  belongs_to :player1, class_name: 'Player'
  belongs_to :player2, class_name: 'Player', optional: true

  # Domain invariants — simple rules
  validate :validate_rules

  def validate_rules
    errors.add(:wins_not_negative, 'Win counts must not be negative') unless (((player1_wins.nil? || player1_wins >= 0) && (player2_wins.nil? || player2_wins >= 0)))
    errors.add(:max_three_games, 'Win counts cannot exceed 2 in a best-of-3 match') unless (((player1_wins.nil? || (player1_wins >= 0 && player1_wins <= 2)) && (player2_wins.nil? || (player2_wins >= 0 && player2_wins <= 2))))
  end

  # Domain invariants — IMPLIES rules
  validate :validate_implies

  def validate_implies
    errors.add(:base, 'BYE match must not have a second player') if (status == 'b_y_e') && !player2.nil?
    errors.add(:base, 'Match end time must be after start time') if (!ended_at.nil?) && !((ended_at.nil? || (!started_at.nil? && ended_at > started_at)))
    errors.add(:base, 'Completed match must have a start time') if (status == 'completed') && started_at.nil?
  end

  def to_s
    table_number.to_s
  end

  # Business operations

  def record_result(p1_wins, p2_wins)
    # TODO: implement record_result
  end

  def finalize_result
    # TODO: implement finalize_result
  end

  def determine_winner
    # TODO: implement determine_winner
    nil
  end

  def concede(player_id)
    # TODO: implement concede
  end

  def draw
    # TODO: implement draw
  end

  # Lifecycle state machine
  ALLOWED_TRANSITIONS = {
    'pending' => ['active', 'b_y_e'],
    'active' => ['completed', 'draw'],
  }.freeze

  def assert_transition!(to_state)
    allowed = ALLOWED_TRANSITIONS.fetch(status.to_s, [])
    unless allowed.include?(to_state.to_s)
      raise ArgumentError, "Transition #{status} -> #{to_state} not allowed"
    end
  end

  def as_json(options = {})
    hash = super(options)
    hash['startedAt'] = hash.delete('started_at') if hash.key?('started_at')
    hash['endedAt'] = hash.delete('ended_at') if hash.key?('ended_at')
    hash
  end
end
